### ISSUE TITLE
OpenACC kernel transformation

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#185]
We want to support kernel transformations within PSyclone. This issue will build the necessary interfaces and implement (some of) them using the fparser2 AST.

**Comment by TeranIvy** [comment]
What do you think by kernel transformations - PSyclone transformations within kernels (hence fparser2)? Thanks!

**Comment by arporter** [comment]
It's a long story. In #155 I was going to do it by obtaining an XCodeML representation of the kernel. However, that requires Omni which in turn requires Java, yacc, lex etc. Rupert and I discussed it yesterday and decided that yes, we would return to fparser2 for this. Initially my target 'transformation' is simply to be able to add `!$acc routine` into a kernel body. However, I'll try to design for the more general case where we may want to re-order loops etc. (possibly via Claw).

**Comment by arporter** [comment]
I have the addition of a directive to a kernel fparser2 AST working. This shows the need for a nice, higher-level API on top of the fparser2 AST. Apart from that, the next step is to manage the creation of a new kernel source file at code generation time.

**Comment by arporter** [comment]
"Code-generation time" == a call to `psy.gen`. Possibly I should add a new, `psy.kernel_gen` method that, for each kernel that has an fparser2 AST, generates a new kernel file. However, that brings the need to re-name kernels which is something that I spent quite a lot of time on in #120. On discussion with Rupert, that can be left to a separate Issue though.

**Comment by arporter** [comment]
I need to finish off the tests but am waiting on #177 because that re-factors some of the test utilities slightly. Besides that, I need to add documentation to the Transformations section and also the Developers' Guide.

**Comment by arporter** [comment]
Travis tests are currently failing because this branch needs features of fparser that are on master but not in a release. I've added documentation for the new transformation but still need to write the associated part of the developers' guide.

**Comment by rupertford** [comment]
#202 is now on master. However, #201 needs to be done before closing this issue.

**Comment by arporter** [comment]
#201 was done a while ago which means this one is all done. Closing issue.
